### PR TITLE
Token aware fix

### PR DESCRIPTION
**Pull request by conorbros**
We only want to select one replica from each rack to use as a potential choice for token aware routing.

### CODE REVIEW

File: shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs [hunk 1]
```diff
@@ -171,12 +172,14 @@ impl NodePool {
...
                 keyspace.replication_factor,
             )
+            .unique()
             .collect::<Vec<uuid::Uuid>>();
 
         let (dc_replicas, rack_replicas) = self
             .nodes
             .iter_mut()
             .filter(|node| replica_host_ids.contains(&node.host_id) && node.is_up)
+            .unique_by(|node| node.rack.clone())
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs:182]: Wouldnt this result in rack_replicas always containing 0 or 1 elements?

Comment (conorbros) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs:182]: Yes

Comment (rukai) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs:182]: In that case should we treat rack_replicas as having one value?
image

**Comment by rukai** [comment]
> We only want to select one replica from each rack to use as a potential choice for token aware routing.

I think we need more justification for this statement.
Is this what other drivers do?
Is it how cassandra token aware routing is designed to work as per some spec?

Like off the top of my head routing between multiple nodes in a single rack seems like a reasonable thing to do?

**Comment by conorbros** [comment]
Yes the other drivers (scylla at least) will select one replica from each rack. I think it makes more sense to do that so if we don't find one in the same rack we are spreading the load better across the other racks.

**Comment by benbromhead** [comment]
How are we placed on resolving this?

**Comment by conorbros** [comment]
Looking into this in the improved routing issue. (best of two random choices algorithm)

**Pull request closed without merging**